File: frontend/src/types/analysis.ts
import type { SimulationRequest } from "./simulation";

export interface ComparisonResultItem {
  scenario_name: string;
  technology_name: string;
  stored_energy_kwh: number;
  delivered_energy_kwh: number;
  round_trip_efficiency: number;
  nominal_power_kw: number;
  initial_capex: number;
  annual_opex: number;
  annual_discharged_energy_mwh: number;
  lcos_per_mwh: number | null;
}

export interface ComparisonResponse {
  success: boolean;
  results: ComparisonResultItem[];
}

export interface SensitivityRequest {
  base_scenario: SimulationRequest;
  parameter_path: string;
  min_val: number;
  max_val: number;
  steps: number;
}

export interface SensitivityResultItem {
  parameter: string;
  value: number;
  lcos: number | null;
  capex: number;
  annual_energy_mwh: number;
}

export interface SensitivityResponse {
  success: boolean;
  parameter: string;
  results: SensitivityResultItem[];
}

export interface MonteCarloRequest {
  base_scenario: SimulationRequest;
  parameter_ranges: Record<string, [number, number]>;
  iterations: number;
  seed?: number | null;
}

export interface MonteCarloResultItem {
  iteration: number;
  sampled_values: Record<string, number>;
  lcos: number | null;
  capex: number;
  annual_energy_mwh: number;
  round_trip_efficiency: number;
}

export interface MonteCarloResponse {
  success: boolean;
  iterations: number;
  results: MonteCarloResultItem[];
}

File: api/schemas/analysis.py
from typing import Dict, List

from pydantic import BaseModel, Field

from api.schemas.scenario import SimulationRequest


class ComparisonRequest(BaseModel):
    scenarios: List[SimulationRequest] = Field(min_length=2)


class ComparisonResultItem(BaseModel):
    scenario_name: str
    technology_name: str
    stored_energy_kwh: float
    required_charge_energy_kwh: float
    delivered_energy_kwh: float
    charge_efficiency: float
    discharge_efficiency: float
    round_trip_efficiency: float
    nominal_power_kw: float
    initial_capex: float
    annual_opex: float
    annual_discharged_energy_mwh: float
    lcos_per_mwh: float | None


class ComparisonResponse(BaseModel):
    success: bool
    results: List[ComparisonResultItem]


class SensitivityRequest(BaseModel):
    base_scenario: SimulationRequest
    parameter_path: str
    min_val: float
    max_val: float
    steps: int = Field(gt=0)


class SensitivityResultItem(BaseModel):
    parameter: str
    value: float
    lcos: float | None
    capex: float
    annual_energy_mwh: float


class SensitivityResponse(BaseModel):
    success: bool
    parameter: str
    results: List[SensitivityResultItem]


class MonteCarloRequest(BaseModel):
    base_scenario: SimulationRequest
    parameter_ranges: Dict[str, tuple[float, float]]
    iterations: int = Field(gt=0)
    seed: int | None = None


class MonteCarloResultItem(BaseModel):
    iteration: int
    sampled_values: Dict[str, float]
    lcos: float | None
    capex: float
    annual_energy_mwh: float
    round_trip_efficiency: float


class MonteCarloResponse(BaseModel):
    success: bool
    iterations: int
    results: List[MonteCarloResultItem]
